### DISCUSSION TITLE
UnusedMethod rule only seems to fire if there's a variable declared/method called inside the class

### DISCUSSION THREAD

**Discussion by sam-gearset** [pmd/pmd#5200] [Q&A]
After playing around most of the morning trying to get this rule (https://pmd.github.io/pmd/pmd_rules_apex_design.html#unusedmethod) to fire and getting very inconsistent behaviour (still struggling in VSCode....), it seems like this rule will only fire if there's a variable declaration or a local method called in the class itself, otherwise it completely ignores it. Can anyone confirm this is expected and if so could we update the documentation to say so?
Thanks,

**Comment by jsotuyod** [comment]
No, it shouldn't. The rule itself relies on [ApexLink](https://github.com/nawforce/ApexLink) that does the proper unused code detection. As long as you have set `PMD_APEX_ROOT_DIRECTORY` properly to a directory containing a complete `sfdx-project.json` file it should work. Bear in mind the metadata needs to be up to date for the rule to work properly, so a build may be required.

If with all of that, you still can't get it to work, we should put together a small reproducer project and open an issue directly with them.

**Reply by sam-gearset** [reply]
Thanks for coming back.
I've trialled it quite extensively in VSCode with both PMD and SF Code Analyzer, with default runs and specific queries using 'pmd check', making sure the env variable is set and the sfdx-project.json, and it only seems to consistently fire if i've used a secondary method to call the 'unused method' in the same class, or declared a variable and returned from that unused method. We've also tested this in Gearset (as they run PMD 7.4.0 on proposed packages) and same behaviour there.
 I can send over some screenshots of my examples if that would help?

**Reply by jsotuyod** [reply]
probably, yes!

**Reply by sam-gearset** [reply]
<img width="503" alt="Screenshot 2024-09-10 at 13 47 40" src="https://github.com/user-attachments/assets/2bdfdacc-3bcc-4b83-856c-e2ed76788222">
<img width="1204" alt="Screenshot 2024-09-10 at 13 47 55" src="https://github.com/user-attachments/assets/e4f2ca4e-972c-4e8a-aa48-76e3239a3041">
<img width="510" alt="Screenshot 2024-09-10 at 13 48 09" src="https://github.com/user-attachments/assets/07fb6161-cb51-4496-a398-72bc00c4b02a">
<img width="1038" alt="Screenshot 2024-09-10 at 13 48 30" src="https://github.com/user-attachments/assets/b9bc7b13-4e36-490a-8cfb-f6a3e3ef58d9">
Got some examples here - if i have a test class with 2 methods that are never called anywhere else, and use pmd check, i'd expect to see the violation for UnusedMethod. However, if i add a third method in calling one of the other 2 methods, only then do i see the UnusedMethod rule fire.
Same example below, but using a private variable declaration instead to force it to trigger
<img width="504" alt="Screenshot 2024-09-10 at 13 50 24" src="https://github.com/user-attachments/assets/f36d2ef7-0436-49aa-bba9-0cf4577dcff0">
<img width="1191" alt="Screenshot 2024-09-10 at 13 50 34" src="https://github.com/user-attachments/assets/9eb8ae03-d9a6-4263-91e7-b7683df635fd">
<img width="538" alt="Screenshot 2024-09-10 at 13 50 48" src="https://github.com/user-attachments/assets/ea5bd885-128f-402e-b3c6-80769a06bdd1">
<img width="1031" alt="Screenshot 2024-09-10 at 13 50 54" src="https://github.com/user-attachments/assets/412e8323-14bf-46cf-a860-4e9d5eeb7c8c">

**Reply by adangel** [reply]
While experimenting myself, I've found the following comment in our test for the rule:
https://github.com/pmd/pmd/blob/469b2a67ae375bde52bb33ccd4fb69e3c2993948/pmd-apex/src/test/resources/net/sourceforge/pmd/lang/apex/rule/design/UnusedMethod/project1/src/Foo.cls#L6-L8

At some point, apex-link will consider the whole class as unused and won't report any method anymore. Not sure, when exactly the whole class is unused (e.g. what are the entry points? the examples we use here are very unrealistic, because you probably implement/extend some apex specific API like a controller which has some public API that will be called by the framework?). I guess, this is what happens and this can be prevented somehow by adding additional methods/fields.

This leads to the question: Should we have here additionally a rule "UnusedClass"? Or change the rule to report both unused methods and classes and rename the rule to "UnusedDeclaration"?

Also note, that a "*.cls-meta.xml" file is mandatory for apex-link to work correctly.

**Reply by adangel** [reply]
Also: ref https://github.com/ChuckJonas/vscode-apex-pmd/issues/157

**Reply by sam-gearset** [reply]
Hey Andreas - apologies for the late reply! 
Thanks for looking into this. I agree it may be worth bundling up into an UnusedDeclaration rule so it can be more accurately determined if an entire class is unused vs particular methods, which could really help in identifying areas for codebase cleanup in a more detailed manner. 
While this is considered, it could be worth updating the [documentation](https://docs.pmd-code.org/latest/pmd_rules_apex_design.html#unusedmethod) to state this current implication?
Thanks,
Sam

**Reply by adangel** [reply]
Sure, documentation improvements are always welcome!

**Reply by sam-gearset** [reply]
Apologies for the very late reply from me again! Created a quick PR [here](https://github.com/pmd/pmd/pull/5354)
